keps/sig-node/compute-device-assignment.md

```diff
@@ -0,0 +1,153 @@
+
+![image](https://user-images.githubusercontent.com/11936386/42943622-e8c676f2-8b80-11e8-8d8f-96a016c6f3b4.png)
+
+### Network Device Plugins
+
+* As a _Device Vendor_, I manufacture network devices.  Even though the Kubelet exposes different interfaces for network setup (CNI), and device lifecycle management (Device Plugin), I can enable consumption of my network devices without requiring communication between my CNI and Device Plugin implementations.  This allows me to independently develop, test, and release my component for network setup, and my component for managing network devices without any dependencies between them. My Device Plugin exposes the set of network devices on the node, but does _not_ perform network setup when `Allocate` is called. Instead, when `CNI_ADD` is called on my CNI plugin, it queries the kubelet's `/pods` endpoint to discover the network devices it should perform network setup for:
+
+![image](https://user-images.githubusercontent.com/11936386/42944024-2fed118e-8b82-11e8-9e45-fd9cd7d78e8c.png)
+
+### Guaranteed Device Scheduling
+
+* As a _Cluster Admin and Application Developer_, I have a workload with strict requirements about the set of devices it is assigned that cannot be expressed with current kubernetes APIs (e.g. numa topology requirements). If my workload is assigned to a node that has the devices I need, but not in the correct configuration, my workloads performance suffers considerably.  Thus, the node's best-effort attempt to provide a "good" set of devices is not enough to meet the needs of my workload.  I can run a custom scheduler to assign devices to containers during scheduling, and can _guarantee_ that the set of devices the workload is assigned meets my requirements.
+
+## Changes
+
+### API Changes
+A new object `ComputeDevice` will be defined and a new variable `ComputeDevices` will be added in `Container` object which will represent a list of `ComputeDevice` objects.
+
+NOTE: `ComputeDevices` will be added as an alpha field and will be behind a feature-gate, `ComputeDevices`  
+
+```golang
+// ComputeDevice describes the devices assigned to this container for a given ResourceName
+type ComputeDevice struct {
+	// DeviceIDs is the list of devices assigned to this container
+	DeviceIDs []string
+	// ResourceName is the name of the compute resource
+	ResourceName string
+}
+
+// Container represents a single container that is expected to be run on the host.
+type Container struct {
+    ...
+	// ComputeDevices contains the devices assigned to this container
+	// This field is alpha-level and is only honored by servers that enable the ComputeDevices feature.
+	// +optional
+	ComputeDevices []ComputeDevice
+	...
+}
+```
+
+To be able to initialize/update `ComputeDevices`, a new verb, `BindResource()`, will be defined for binding `ResourceBinding` objects with a pod object:
```

Comment:
I have not considered it, but I am open to the idea.  What are the advantages of using a CRD?
At first glance, it seems odd to me to have a CRD be used by core components (kubelet).  It doesn't seem to fit with the [description of a CRD in our docs](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) either:
`A custom resource is an extension of the Kubernetes API that is not necessarily available on every Kubernetes cluster. In other words, it represents a customization of a particular Kubernetes installation.`

Reply:
Alpha fields seem to be having issues still (@thockin was talking about it earlier today). One of Tim's ideas was to create a CRD per pod that uses extended devices with the same name and namespace that includes device properties like Device ID.

Reply:
`CRD per pod that uses extended devices` -- What does it mean by CRD "uses" extended resources?
Just thinking loud, CRD object being proposed to create just to hold some pod metadata.

@vish can you please elaborate a bit more on what a CRD will look like and what it would hold?

Reply:
@edwarnicke without init container how pod-id will be be updated at CRD object? How monitoring agent will figure out for a particular pod which CRD object is associated?

Reply:
i am not following the CRD approach.  this introduces another resource that the kubelet needs to list/watch and update.  we are already discussing a number of other resources that the kubelet would list/watch update, that i am starting to get concerned we will hit scale concerns.

Reply:
Here's a thread on sig-arch on a closely related topic - https://groups.google.com/forum/#!msg/kubernetes-sig-architecture/y9FulL9Uq6A/mCEKZpLyAQAJ

@derekwaynecarr the main question to answer is how do we add fields to existing v1 API objects with sub-field versioning. Alpha fields are the current guidance, but they have some limitations. shadow CRDs seem to work for use cases. A recent example is health checks.
@edwarnicke point was that a CRD might provide leeway to introduce more introspection fields in the future with minimal overhead.
I do see the drawbacks of kubelet writing to more and more objects. 
I think one question that might help make decision here is if we foresee more attributes being exposed in addition to device ID in the future from the kubelet (or via it).

Reply:
@vishh One other thing to keep in mind... it sounds like we are dealing with a mix of two things:

* Checkpointing for Kubelet
* Operational Data for Monitoring things

Using the Pod.spec for checkpointing, while not unprecedented, does feel a bit wrong to me.  Having a separate CRD also allows us to express what is essentially operational state more clearly.   For example it could allow Kubelet to expressed checkpointing information and real operational status clearly.  It also allows this to be a properly alpha thing, while we all gain operational experience with it and inevitably realize it needs to evolve as we learn.

Reply:
@edwarnicke seems you missed my query above, https://github.com/kubernetes/community/pull/2454#discussion_r211661982 :)